### ISSUE TITLE
(inactive) ctgPHYL

### ISSUE THREAD

**Issue by bicbioeng1** [Bioconductor/Contributions#1353]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/bicbioeng/ctgPHYL

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by mtmorgan** [comment]
looks like this package could be a great addition, but it needs to interface with standard Bioconductor single-cell representations, e.g., [SingleCellExperiment][1] and [TreeSingleCellExperiment][2]

[1]: https://bioconductor.org/packages/SummarizedExperiment
[2]: https://bioconductor.org/packages/TreeSummarizedExperiment

**Comment by bicbioeng1** [comment]
Thank you! We will get to work on interfacing with the single-cell classes
and post an update when ready.

**Comment by bicbioeng1** [comment]
Dear Mr. Morgan,

Thank you so much for your insight for our package submission.
After discussing your suggestion to interface with other Bioconductor
single-cell structures and doing some testing, our team is not certain
these representations are best for the ctgPHYL package core goal and future
development. Our data inputs are tied directly to the cellscape, timescape,
and mapscape packages, and our workflow generates graph-based structure
(igraph compatible)  with the intention of broad downstream analysis
leveraging systems biology tools such as Cytoscape. Since the igraph format
interfaces well with Cytoscape, we would like to preserve the structure of
our ctgPHYL class as best as possible. If you have suggestions on how to
interface with the other single-cell representations while preserving this
aspect of our package, please let us know.

Thank you for your time!

**Comment by hpages** [comment]
Hi,

Thanks for this submission. Here is some feedback about the ctgPHYL package:

1. Wouldn't it be simpler to have packages cellscape, timescape, mapscape, htmlwidgets in `Imports` (or `Depends`) rather than in `Suggests`. ctgPHYL heavily relies on these packages for almost everything it does so it's not really an option for the users to not have them installed if they're to use ctgPHYL. One immediate benefit of having these packages in `Imports` (or `Depends`) is that the user then no longer needs to install them separately so section 2 of your vignette becomes **much** simpler. Also you no longer need to do things like `if (!requireNamespace("mapscape", quietly = TRUE)) ...` internally.

2. ctgPHYLset objects display their plots in the browser in one tab per plot. Unfortunately the names of the tabs are cryptic (they are the path to a temporary HTML file e.g. `file:///tmp/Rtmphv10IV/viewhtmlf966baf227e/index.html`) so not helpful. I would suggest using more useful names e.g. display the tree type (`cellscape`, `timescape`, or `mapscape`). This would make it easier for the user to quickly identify tabs when switching between many tags.

3. IIUC the benefit of using ctgPHYL over using cellscape, timescape, and mapscape directly is to allow bundling 3 types of data (cellscape, timescape, and mapscape data) together in a single object and to make it easy to plot these data. However, I have to say that I don't find using ctgPHYL particularly easier than using cellscape, timescape, and mapscape directly. I would even argue that using the former seems less convenient and more error prone.
    For example, the user needs to do a lot of things to create a fully operational ctgPHYL object: 4 calls to the `cellscapeData()` setter + 4 to the `timescapeData()` setter + 6 to the `mapscapeData()` setter. These calls don't seem to perform any check e.g. in case of a typo in the 2nd argument of a setter (e.g. `cellscapeData(toyPHYLset, "tree_edge")` instead of `cellscapeData(toyPHYLset, "tree_edges")`) nothing is said on the moment but the resulting object is invalid and will fail later in an obscure way. Furthermore the user then needs to call `generate_tree()` 3 times before they can actually display the plots. So in the end they need to do more work than if they had called `cellscape()`, `timescape()`, and `mapscape()` directly on their data and, while doing so, they're more likely to make mistakes and these mistakes will silently be ignored.

4. Also it is not clear what the benefit is to bundle the 3 types of data together in a single object versus using 3 separate objects (one for each type of data). Are the 3 data components stored in a ctgPHYL object correlated in any way? If so, do you check that this correlation is actually present when the user adds the data to the object and do you take advantage of this correlation somewhere?

5. Why wouldn't the `cellscapeData()`, `timescapeData()`, and `mapscapeData()` getters support the 2nd argument like the setters do?
    ```
    cellscapeData(toyPHYLset, "tree_edges")
    # Error in cellscapeData(toyPHYLset, "tree_edges") : 
    #   unused argument ("tree_edges")
    ```

6. Surprisingly the documentation never says that ctgPHYL is for visualization of **phylogenetic** data. The name of the package kind of suggests this but it should be clearly stated in the `Title` and `Description` fields of the `DESCRIPTION` file and in the Introduction of the vignette. Also please add the term `Phylogenetics` to your `biocViews` field (this is a new term that is only available in the very latest version of the biocViews package i.e. in version 1.55.6).

7. Both the `Description` field in the `DESCRIPTION` file and the Introduction in the vignette are too vague. The terms used are too generic (e.g. "a simplified way to build trees") which doesn't really help. It's only after spending some time with the package that I really understood what it's trying to achieve. Please be a lot more specific and provide more details in both places.

Thanks

**Comment by hpages** [comment]
Hi @bicbioeng1 ,

Do you intend to follow up on this submission?

Thanks,
H.

**Comment by bicbioeng1** [comment]
Hi Hervé,

We apologize for the delay, we had put ctgPHYL on hold for some time but we
will send an update very soon. Thank you for taking the time to review our
package,
your comments were very helpful!

Best,
Evgeni

**Comment by bicbioeng1** [comment]
Apologies for the delay! Here are answers your questions and comments,
including updates to the package:

   1.

   Wouldn't it be simpler to have packages cellscape, timescape, mapscape,
   htmlwidgets in Imports (or Depends) rather than in Suggests. ctgPHYL
   heavily relies on these packages for almost everything it does so it's not
   really an option for the users to not have them installed if they're to use
   ctgPHYL. One immediate benefit of having these packages in Imports (or
   Depends) is that the user then no longer needs to install them separately
   so section 2 of your vignette becomes much simpler. Also you no longer
   need to do things like if (!requireNamespace("mapscape", quietly = TRUE))
   ... internally.

Changed the 'Imports' field of our DESCRIPTION file to include the workflow
packages.

   1.

   ctgPHYLset objects display their plots in the browser in one tab per
   plot. Unfortunately the names of the tabs are cryptic (they are the path to
   a temporary HTML file e.g.
   file:///tmp/Rtmphv10IV/viewhtmlf966baf227e/index.html) so not helpful. I
   would suggest using more useful names e.g. display the tree type
   (cellscape, timescape, or mapscape). This would make it easier for the user
   to quickly identify tabs when switching between many tags.

Was not able to locate the cryptic tempfile tab names. The HTML widget file
names include the package being used (cellscape, timescape, or mapscape),
and the names are included in both the ctgPHYLset@originalTrees slot and
the html files saved in the CTG-Output/Plots folder. Are these tab names
displayed in a web browser, or in Rstudio?

   1.

   IIUC the benefit of using ctgPHYL over using cellscape, timescape, and
   mapscape directly is to allow bundling 3 types of data (cellscape,
   timescape, and mapscape data) together in a single object and to make it
   easy to plot these data. However, I have to say that I don't find using
   ctgPHYL particularly easier than using cellscape, timescape, and mapscape
   directly. I would even argue that using the former seems less convenient
   and more error prone.
   For example, the user needs to do a lot of things to create a fully
   operational ctgPHYL object: 4 calls to the cellscapeData() setter + 4 to
   the timescapeData() setter + 6 to the mapscapeData() setter. These calls
   don't seem to perform any check e.g. in case of a typo in the 2nd argument
   of a setter (e.g. cellscapeData(toyPHYLset, "tree_edge") instead of
   cellscapeData(toyPHYLset, "tree_edges")) nothing is said on the moment but
   the resulting object is invalid and will fail later in an obscure way.
   Furthermore the user then needs to call generate_tree() 3 times before they
   can actually display the plots. So in the end they need to do more work
   than if they had called cellscape(), timescape(), and mapscape() directly
   on their data and, while doing so, they're more likely to make mistakes and
   these mistakes will silently be ignored.

Added some error-checking to prevent obscure failures in the workflow.

In this version, the aim for ctgPHYL was for users to enter all their data
and have the rest of the workflow of their desired package be automated. We
will work on the followup version to improve the ease-of-use when it comes
to data input and tree generation.

   1.

   Also it is not clear what the benefit is to bundle the 3 types of data
   together in a single object versus using 3 separate objects (one for each
   type of data). Are the 3 data components stored in a ctgPHYL object
   correlated in any way? If so, do you check that this correlation is
   actually present when the user adds the data to the object and do you take
   advantage of this correlation somewhere?

Although one purpose of our package is to streamline the process of
visualizing cell

phylogeny with various R packages (at this time the -scape packages),
another

important feature is transforming each unique tree output into a simplified
igraph format

- the SIF output file. These SIF files provide the community with
generalizable FAIR-compliant tools for cell interaction output, allowing
downstream analysis with network visualization tools like Cytoscape. Our
intent was to provide a tool for users to visualize data using existing R
packages and generate a useful output for further analysis. As this is not
emphasized very well in the current documentation, we've made some updates
to reflect these features.


   1.

   Why wouldn't the cellscapeData(), timescapeData(), and mapscapeData()
   getters support the 2nd argument like the setters do?
   cellscapeData(toyPHYLset, "tree_edges")

# Error in cellscapeData(toyPHYLset, "tree_edges") :

#   unused argument ("tree_edges")

Changed the setters to match the getters.

   1.

   Surprisingly the documentation never says that ctgPHYL is for
   visualization of phylogenetic data. The name of the package kind of
   suggests this but it should be clearly stated in the Title and Description
   fields of the DESCRIPTION file and in the Introduction of the vignette.
   Also please add the term Phylogenetics to your biocViews field (this is a
   new term that is only available in the very latest version of the biocViews
   package i.e. in version 1.55.6).

We included Phylogenetics in our BiocViews, and let users know ctgPHYL is
for phylogenetic data.

   1.

   Both the Description field in the DESCRIPTION file and the Introduction
   in the vignette are too vague. The terms used are too generic (e.g. "a
   simplified way to build trees") which doesn't really help. It's only after
   spending some time with the package that I really understood what it's
   trying to achieve. Please be a lot more specific and provide more details
   in both places.

Updated the documentation for both the package DESCRIPTION and the vignette
in order to better reflect the purpose of our package.

In addition to these changes, we added the option for users to set their
output directory in the generate_tree() function. All changes have been
pushed to the Bioconductor repository. Thanks again, we appreciate your
time and your feedback!

Best,

Evgeni

**Comment by hpages** [comment]
Hi Evgeni

Thanks for the improvements.

Your last comment above has formatting issues that make it hard to read. Please use the _Preview_ feature to check the formatting of your comment before you post it. Thanks!

Unfortunately my concerns about usability remain.

**About item 2:**

> Are these tab names displayed in a web browser, or in Rstudio?

Yes, please see item 2 in my original feedback. The fact that the tabs in the browser have long and cryptic names (e.g. `file:///tmp/Rtmphv10IV/viewhtmlf966baf227e/index.html`) are a serious usability issue. Basically I don't know where my plots are because I have no idea what plot is in what tab, unless I click on the tab. But with many opened tabs, that is very impractical.

**About item 3:**

The fact that you still need 14 calls to the `cell/time/mapscapeData()` setters to stuff a ctgPHYLset object is also a usability concern that I was hoping you would tackle. Also, as I reported, it seems unnecessary to require the user to call `generate_tree()` 3 times before the show() method for ctgPHYLset objects will draw the plots in the browser. In other words you need to think about how to improve the way your users will interact with their ctgPHYLset object and make it less cumbersome.

**New item:** (item 8)

The Description field in the DESCRIPTION file and the vignette's introduction now say:

> ... and provides a unified output format for integration with downstream data
> analysis workflows and Cytoscape.

It would be great to backup this claim with an example.

Thanks,
H.

**Comment by bicbioeng1** [comment]
Hervé,

*About Item 2:*
I wasn't sure if I could update the package in time, but I would like to
try and address your concerns. I'm still not sure what browser tabs you are
referring to -
ones in the RStudio Viewer, ones in a browser like Chrome, or somewhere
else? I just can't seem to locate the cryptic filenames.

*About Item 3:*
In terms of usability, I have some updates that will allow users to
provide a list of parameters to cellscape(),mapscape(), and timescape()
instead of one
parameter at a time. Do you think that would improve usability enough? I'm
not sure how to make the show() method work with only 1 plot, I would have
to look into that.

 *About Item 8*
We can do that, although I don't think so by tomorrow. Does that mean
further changes would have to wait until the next development cycle is over?

Thanks again,
Evgeni

**Comment by hpages** [comment]
Unfortunately it is too late for inclusion in BioC 3.11. Let's focus on improving usability of the package before its inclusion to Bioconductor. Once a package is accepted and people start using it, it's much harder to make changes to the API without breaking people's code. If everything goes well, the package will be added to BioC 3.12 (normally to be released in fall this year). Of course the package would be added to BioC 3.12 before that, i.e. any time between the BioC 3.11 release (next Tuesday) and the BioC 3.12 release.

BTW you or your lab also [submitted the ctgGEM package](https://github.com/Bioconductor/Contributions/issues/1334) recently and the package got accepted. However, the package is currently failing on all platforms. See build report [here](https://bioconductor.org/checkResults/3.11/bioc-LATEST/ctgGEM/). Apparently the culprit is the cellTree package that they rely on. I've already contacted the maintainer of the ctgGEM twice by email (maintainer is `USD Biomedical Engineering <bicbioeng@gmail.com>`) . They need to be more proactive about fixing the problem. Otherwise ctgGEM won't make it into the BioC 3.11 release either.

Thanks,
H.

**Comment by bicbioeng1** [comment]
Thank you for the update, we'll make sure to improve ctgPHYL usability for
the upcoming release.

As for ctgGEM, we didn't realize the daily builds happened around 2 pm CT,
so the package updates were not checked yesterday as expected. Hopefully
that won't be an issue after today's build.

**Comment by lshep** [comment]
Do you plan on making updates to your package in a timelly manner.  Packages not actively making changes and participating in the review process will be closed for inactivity and you can request it be opened when you have time to work through the submission process?

**Comment by bicbioeng1** [comment]
We will need some time to update our package. If we close it temporarily,
could we request via email to have it reopened for submission? Or will we
have to restart the submission process from the start?

**Comment by lshep** [comment]
If we close it temporarily,  when ready just leave a message on this issue that you would like it reopened and the process to resume.  
Cheers,
